File: modules/cadete-background-tracking/android/src/main/java/expo/modules/cadetebackgroundtracking/CadeteBackgroundTrackingModule.kt
package expo.modules.cadetebackgroundtracking

import android.content.Context
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class CadeteBackgroundTrackingModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("CadeteBackgroundTracking")

    AsyncFunction("startTracking") { userId: String, meteorUrl: String? ->
      val context = requireContext()
      CadeteTrackingService.start(context, userId, meteorUrl)
      CadeteTrackingService.getStatus(context)
    }

    AsyncFunction("stopTracking") {
      val context = requireContext()
      CadeteTrackingService.stop(context)
      CadeteTrackingService.getStatus(context)
    }

    AsyncFunction("syncTracking") { enabled: Boolean, userId: String?, meteorUrl: String? ->
      val context = requireContext()
      CadeteTrackingService.sync(context, enabled, userId, meteorUrl)
      CadeteTrackingService.getStatus(context)
    }

    AsyncFunction("getStatus") {
      CadeteTrackingService.getStatus(requireContext())
    }
  }

  private fun requireContext(): Context {
    return requireNotNull(appContext.reactContext?.applicationContext) {
      "CadeteBackgroundTracking necesita un applicationContext disponible"
    }
  }
}

File: modules/cadete-background-tracking/ios/CadeteBackgroundTrackingModule.swift
import ExpoModulesCore

public final class CadeteBackgroundTrackingModule: Module {
  public func definition() -> ModuleDefinition {
    Name("CadeteBackgroundTracking")

    AsyncFunction("startTracking") { (userId: String, meteorUrl: String?) -> [String: Any?] in
      return CadeteNativeLocationService.shared.startTracking(userId: userId, meteorUrl: meteorUrl)
    }

    AsyncFunction("stopTracking") { () -> [String: Any?] in
      return CadeteNativeLocationService.shared.stopTracking(clearConfig: true)
    }

    AsyncFunction("syncTracking") { (enabled: Bool, userId: String?, meteorUrl: String?) -> [String: Any?] in
      return CadeteNativeLocationService.shared.syncTracking(enabled: enabled, userId: userId, meteorUrl: meteorUrl)
    }

    AsyncFunction("getStatus") { () -> [String: Any?] in
      return CadeteNativeLocationService.shared.currentStatus()
    }
  }
}
